Feature/initial mm analytics

brecriffs opened the pull request:
## Description

- Created a docker-compose to run containers for the metabase and mysql images
- Created a Makefile to allow developers to use similar commands as to what we are used to
- Added a README to provide instructions to developers on how to run the software
- Created a script in the /bin directory which restores sql backups to the mysql container's local volume, for use by metabase

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This goal was to create a new repository with all the configuration, documentation, and scripts
to let another developer run metabase and a database locally and produce the defined analytic
reports.

## How has this been tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested locally and was able to clone repo and run metabase

## Types of changes
<!--- What types of changes does your code introduce? -->
<!--- Please remove all lines which don't apply. -->
- ✅ New feature (non-breaking change which adds functionality)

## Checklist:
<!--- Please go over all the following points. -->
<!--- Again, remove any lines which don't apply. -->
<!--- Pull Requests that don't fulfill all [REQUIRED] requisites are likely -->
<!--- to be sent back to you for correction or will be rejected.  -->
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ My change requires a change to the documentation and I have updated it accordingly.

brecriffs committed:
QA and improve queries

brecriffs committed:
Update channels query

brecriffs committed:
Final QA of queries and optimization of visualisations

brecriffs committed:
update makefile

brecriffs committed:
Initial commit. Testing to see if I can clone repo and run commands to start metabase and see reports

brecriffs committed:
Update README

brecriffs committed:
Update README

brecriffs committed:
Update README.md

brecriffs committed:
Add database-backups directory with gitignore

mlippert committed:
Add github Issue and Pull Request templates

The fact that the 1st PR created didn't have any description shows that these are needed!

mlippert commented:
@brecriffs Umm, this PR needs a description. I just added a commit w/ some templates for issues and PRs created in the future, but you should still describe what this PR was for.

mlippert commented on Makefile:
```diff
@@ -0,0 +1,25 @@
+
+#runs the docker-compose.yml showing logs
+up :
+	@echo "Starting Riff Business Analytics.";
```
Why the trailing semicolon? (on this line and many others)

brecriffs replied:
Removed semi-colons. Was unclear about the syntax required.

mlippert replied:
Don't stay unclear. If you don't understand the syntax, find and read the docs, don't guess. https://www.gnu.org/software/make/manual/

mlippert commented on Makefile:
```diff
@@ -0,0 +1,25 @@
+#runs the script to restore database backups for metabase to run queries on
+restore :
+	@echo "Restoring mysql database backups from 'database-backups' folder.";
+	@bash bin/restore-database-backups.sh;
```
why are you explicitly executing bash here?

brecriffs replied:
> Using the `Makefile` in riff-docker as an example, having the default target be help which tells you what the useful targets in the Makefile are and what they do is worth doing.
> 
> I still have to actually try to run the environment and see if I can create any charts, following the documentation. I may have more comments after I do that.

Added help as default target.

brecriffs replied:
Without explicitly using @bash, I get a permission denied error...Not exactly sure what the root issue is.

mlippert replied:
Ah, did you try running this script from your terminal directly?
I suspect you haven't made the file executable (`chmod +x bin/restore-database-backups.sh`)

brecriffs replied:
Also, it made the command silent I guess, which I thought made the console look cleaner.

tautropfli replied:
🤖 Beep boop. This is the totally-automated-I-am-definitely-not-a-human @bash bot saying hello 👋. Beeb boop 🤖.

jaedoucette replied:
+1 on permissions. Could also use `1>/dev/null` if command silence is desired, to preserve stderr output?

brecriffs replied:
Ok I will try that. Command silence just seemed sleeker to me, but I don't think it's a priority.

mlippert replied:
just a note. command silence from the command and using a subshell are 2 different things.
When you say command silence what exactly do you mean, and at least initially it would seem to me that more feedback is better in case things go wrong in ways you didn't anticipate. Squelching the output seems better when you gracefully handle things going wrong. I just don't think this is at that point yet.

brecriffs replied:
Alright that does make sense. I will leave it for now.

mlippert commented on bin/restore-database-backups.sh:
```diff
@@ -0,0 +1,14 @@
+#
```
It's a good idea for scripts to start w/ a "shebang" '#!' that specified what executable should run the script in the file. So for a bash script start with
```
#! /usr/bin/env bash
```
or
```
#! /bin/bash
```
I've started using the 1st one, but see https://stackoverflow.com/questions/10376206/what-is-the-preferred-bash-shebang

brecriffs replied:
Added shebang

mlippert commented on database-backups/.gitignore:
```diff
@@ -0,0 +1,5 @@
+*
+!.gitignore
+!*.gpg
+!*.sh
```
Are you expecting to add scripts (`*.sh`) to this directory? I could see possibly checking in encrypted files (`*.gpg`) although for backups they're probably too big to want them in the repo.

brecriffs replied:
I just added this from the snippet you sent over slack. I guess that was just an example? I thought the idea was to possibly add the .sql backups here, but only if they were encrypted. So I guess for now, ignore everything in this folder except for the gitignore?

mlippert replied:
I'm pretty sure I said "here is an example".
So I was asking for confirmation that you thought about what was in the gitignore when you added it (or not), and the reasoning behind what you did.
If I had said, here use this, I would expect you to look at it and understand what it does, or otherwise to ask the same questions I just asked you.
Potentially allowing encrypted backups to be in this folder is very reasonable, and a good reason to keep the `!*.gpg`, although if you do that, you should probably not try to restore those files in your restore script.

brecriffs replied:
Ok makes sense.

mlippert commented on bin/restore-database-backups.sh:
```diff
@@ -0,0 +1,14 @@
+   cat "$1" | docker exec -i mysql-business-analytics /usr/bin/mysql --user=root --password=root
+}
+
+for i in database-backups/*
```
perhaps you should restrict the files you loop over to only `*.sql` files?
Also if there are several backups of the same mysql database in the `database-backups` directory what happens? (I'm just wondering if attempting to restore **every** backup file in the directory is going to have the desired result.

brecriffs replied:
Okay made that restriction. I tested with 3 of the same backup, and it just overwrites it, which I think would be the desirable? If you're uploading two backups with a database of the same name, I think that is a user mistake.

mlippert replied:
So I guess I'm wondering about what backups w/ different databases we have that you are anticipating being restored.
I'm not suggesting you change this working code, just curious what future scenarios you were planning for. If you hadn't really given it that much thought, I'm just pointing out that it is worth thinking about as you're doing something like this.

brecriffs replied:
I guess I was anticipating there would be more than one client using our mattermost integration (edu), and that we could use different metabase dashboards for each client's database backup.

mlippert replied:
I see, that makes sense, except you didn't think about the fact that the database name for every client is the same, so they will overwrite one another. But that is perhaps a future enhancement to for this repo to deal with.

brecriffs replied:
I did consider that, and knew the caveat was that the database names would have to be different.  Perhaps this should handle that gracefully and alternatively name the databases at run-time, or at the very least not overwrite them. I guess, like you said, this might be a future enhancement. 

Also, I think during my initial approach, my thinking was that this made things dynamic so I didn't have to hard-code a filename for the .sql file to import.

mlippert replied:
>  this made things dynamic so I didn't have to hard-code a filename
That is true and makes a lot of sense.

mlippert commented on docker-compose.yml:
```diff
@@ -0,0 +1,31 @@
+
+  target-db:
+    image: mysql:5.7
+    container_name: mysql-business-analytics
```
why is there a `container_name` instead of just letting docker-compose create the container name from the directory and the service name?

brecriffs replied:
I just thought it was nice to provide a nice container name, the generated one seemed somewhat cryptic. And I need to refer to the container name inside the script in /bin to restore the sql backups.

mlippert replied:
Those are both an interesting and reasonable points.

However, I don't think we should care how cryptic the container name looks, as long as it is identifiable as to what it is.

As for referring to the container from inside your script, that is a reasonable initial approach. I think longer term we will run into issues doing it that way, but it does solve an otherwise tricky problem. Since you've followed my other naming suggestions for the services, I would suggest at least setting the container name to match that convention. (And add a comment as to why you're setting the container name)

brecriffs replied:
Removed container name from the metabase container since I don't reference it explicitly. Changed the mysql container name to match conventions and added a comment explaining where the name is used.

mlippert commented on docker-compose.yml:
```diff
@@ -0,0 +1,31 @@
+version: '3.6'
+services:
+
+  target-db:
```
I've found that good service names have a short prefix (e.g. `pfm-`, `edu-`) and then a relatively short descriptive name, so perhaps for this service something like `ba-mmdb` or something else that makes sense to you.

brecriffs replied:
Okay makes sense, I've added consistent names to the services.

mlippert commented on docker-compose.yml:
```diff
@@ -0,0 +1,31 @@
+    volumes:
+      - target-db-data:/var/lib/mysql
+
+  metabase:
```
Going along w/ the above comment about service names, perhaps `ba-metabase`. Having a common prefix for the services defined in a compose file that should all work together is helpful in identifying them in the lists of images/containers.

brecriffs replied:
Added consistent names to the services.

mlippert commented on docker-compose.yml:
```diff
@@ -0,0 +1,31 @@
+      - target-db
+
+volumes:
+  target-db-data:
```
A similar comment about volume names as I made above about service names, so keep the prefix you used for the service names, and relate the volume to the service it will be bound to, e.g. `ba-mmdb-data`

brecriffs replied:
Also makes sense, added consistent names to the volumes.

mlippert commented on docker-compose.yml:
```diff
@@ -0,0 +1,31 @@
+
+volumes:
+  target-db-data:
+  metabase-db:
```
similarly this one might be `ba-metabase-db`, just adding the prefix to help when looking at the lists of docker volumes.
Although now that I look more closely I see this volume is not actually referenced by any service.

brecriffs replied:
Ah you're right. I guess since I'm just binding the local folder to the folder inside the metabase container, a volume is never created/needed...and I'm not pointing to it anyways.

mlippert replied:
Right.
Now whether you should be binding to a local folder is a different question.

That will depend on the instructions for getting this up and running.
Ideally, you can export the questions/dasboard into some file that the setup instructions will explain how to import, so you can use a named volume and not have to put the
metabase database into the repo. I'm not sure where you've gotten with that.

brecriffs replied:
It isn't clear to me why that is ideal. What are the downsides to having the metabase database in the repo?

mlippert replied:
It's opaque, it changes when used, so you don't know if you need to commit it again, it can't be merged (because it's an opaque binary file).

brecriffs replied:
Ah I see. Yes, it does change. But unless you change a question or anything on the dashboard, you shouldn't commit it.

brecriffs replied:
I still think it's a better option than having to setup and configure the entire metabase app including user info, database configurations, the dashboard, and each question in the dashboard (and configuring the dynamic variables in each question).

brecriffs replied:
It's possible the dynamic variables get output in the question export. This might be something I should test locally now. It still sounds like a lot of work to just get something up and running locally.

mlippert replied:
The available options for export/import may indeed be more trouble than they're worth at this time, it would be better for many reasons, but may not be worth the effort, the question is what is the effort. And the available options may have addition downsides, again just questions to be answered in order to come to a reasoned decision.

brecriffs replied:
Fair enough. Alright I will look into question export options and get back to you later today.

jaedoucette replied:
+1 on not putting the metabase DB in the repo. 

Relatedly, does the metabase DB contain PII? I think it does. In that case, it is also a bad practice for us to store it unencrypted on local machines.

brecriffs replied:
It doesn't store anything from the mattermost db. Just run-time data...the sql queries (questions), dashboard, and general configuration of metabase.

jaedoucette replied:
Ah, okay. I'm less concerned about it then.

brecriffs replied:
It is admittedly a little awkward to have checked in, partially due to the opaqueness of the code. As you have this software running locally, the metabase DB file will change. But I recommend not committing it unless you make a change to the metabase config, questions, dashboard layouts, etc. This definitely seems easier and more straightforward to me than having to configure metabase every time you run this software. This would involve setting up a user account, configuring the database connection, importing each question and configuring its dynamic variables, and possibly organising a dashboard for ease of use. Having to do that each time seems inefficient. 

As a second note, it is recommended by the metabase community, when moving to a production environment, to abandon the use of the H2 database and migrate to a postgres/mysql database to store the metabase database.

mlippert replied:
That's something to look into going forward. Configure metabase to use the mysql database we're running anyway, and commit the backup of that database instead of the binary database file.
The backup of a mysql database is a text file, and therefore can be diffed, and possibly merged, although that is still likely to be difficult.
There may be reasons that won't work, but it's worth keeping in mind and exploring at some point.

mlippert commented on metabase-db/metabase.db/metabase.db.trace.db:
```diff
@@ -0,0 +1,170 @@
+2019-04-26 21:21:14 jdbc[4]: exception
```
This does not seem like a useful file to have checked in. I do realize that is a file created by the metabase container.

brecriffs replied:
This is a debug log. I agree that it should not be included in the repo. I will add a gitignore to this folder and ignore this file.

mlippert reviewed (commented):
Using the `Makefile` in riff-docker as an example, having the default target be help which tells you what the useful targets in the Makefile are and what they do is worth doing.

I still have to actually try to run the environment and see if I can create any charts, following the documentation. I may have more comments after I do that.

mlippert commented on Makefile:
```diff
@@ -0,0 +1,25 @@
+restore :
+	@echo "Restoring mysql database backups from 'database-backups' folder.";
+	@bash bin/restore-database-backups.sh;
+	@echo "Processes Complete. Metabase running on port 3000.";
```
I'm not sure what you mean by `Processes Complete` or how you can say for sure that metabase is running on port 3000 when you didn't do anything for this target to make that happen.

brecriffs replied:
Sorry that should have been removed from this target. The command used to be up-restore, which did both run docker-compose up, then do the DB restorations. But I found that it sometimes failed because the mysql container wasn't ready. I felt for the scope of the current task, it would be fine to just run `make up`, then after a few minutes, run `make restore` in a new tab. (Which the new tab part needs to be added to the documentation). I think there needs to be a more elegant way of determining when docker-compose finishes so we can @echo something on the screen, but as for the DB restorations, it seems to be successful in waiting for all sql backups to be restored, before hitting the final @echo 'Processes Complete'. Maybe a more descriptive line like 'Database restorations complete.', would be better?

mlippert replied:
I'm not sure that `make restore` does a lot for you. Perhaps the doc should just tell you to run `bin/restore-database-backups.sh`.

Is the echo after the actual work really helpful for all of those targets? If the answer is yes, then great, but if not perhaps it should be removed.

As a 1 time thing you also could just ask that the user supply the container name as an argument to the script, and then you wouldn't have to name it in the docker-compose (although having a constant name does make things easier, it is only an occasional thing).

brecriffs replied:
Removed `make restore` and updated docs accordingly.

mlippert committed:
update README w/ some links and minor formatting

mlippert commented on README.md:
````diff
@@ -1,2 +1,88 @@
+```
+make restore
+```
+Once the `make up` command has completed (about 2-3 minutes), run this command. It will restore databases from all .sql files in the folder ./database-backups
````
It's sort of hard to run `make restore` when `make up` is attached to the terminal, your instructions should therefore probably mention that you'll have to use another terminal to do that.

brecriffs replied:
Added that to the documentation.

mlippert commented on README.md:
```diff
@@ -1,2 +1,88 @@
+
+## View the Dashboard
+
+Once you navigate to port 3000, you will be shown the metabase homepage. I have pinned the Mattermost dashboard to the 
```
'navigate to port 3000' really isn't clear. To be clear you should say something about go to `http://localhost:3000`.
I just did this and did **NOT** see the metabase homepage you describe. What I get is a metabase login page asking for an email address and a password.
I was wondering about the idea of checking the metabase database into the repository and this is making me think that the instructions should include (or maybe just refer to the metabase setup instructions).
A quick google found this discussion of import/export of questions and dashboards, which sounds like it might be a thing to keep in the repo than a pre-setup database. Worth exploring anyway, before having to add an email and password to these instructions.

brecriffs replied:
You are correct, I should perhaps to the documentation that you need to obtain the username and password from the dev team. From looking into it, it seems you can export questions, but not entire dashboards, which seems like it would take a long time. I feel like checking in the database and logging in is the best option.

mlippert reviewed (changes requested):
So the documentation isn't correct at this point. I suspect @brecriffs that you have cookies stored in your browser and that's why you didn't run into the issues I just encountered.
Obviously the documentation needs to be updated to work for someone starting fresh. So I'll wait for updates.

brecriffs commented:
Thanks for the feedback. I think some issues are due to my ignorance relating to makefiles and docker. Will comment on each to keep track of progress for myself and push another commit once complete.

mlippert commented:
@brecriffs Of course, and that's why we do code reviews :smile: . Comment again once you think you've addressed the things you want to and are ready for me to review again.

First you should give the PR a description, even if you decide you need to edit it more afterwards.

brecriffs commented:
> @brecriffs Umm, this PR needs a description. I just added a commit w/ some templates for issues and PRs created in the future, but you should still describe what this PR was for.

Added description. Will revise after addressing other issues if needed.

mlippert commented:
Thanks for updating the description of this PR. Looks good.

brecriffs committed:
Update README.md

brecriffs committed:
Addressed requested changes in PR by Mike

1. Made a help target as the default target in the Makefile
2. Added shebang to the top of the ./bin/restore-database-backups.sh script
3. Modified gitignore in ./database-backups to ignore everything except the gitignore and any encrypted files
4. Restricted the loop in ./bin/restore-database-backups.sh to look for only .sql files instead of all files
5. Added more conventional container name to mysql container, and removed container name from metabase container in docker-compose
6. Added more conventional service names in docker-compose
7. Added more conventional name to the mysql volume and removed the unused metabase volume in docker-compose
8. Added gitignore to ./metabase-db to ignore the metabase.db.trace.db file, which is the metabase debug log
9. Removed unnecessary trailing echos in targets in the Makefile
10. Removed the unnecessary restore target in the Makefile and adjusted the docs to instruct the user how to run the restore script manually
11. Modified docs to instruct user to ask dev team for login creds for metabase
12. Removed unnecessary semicolons from Makefile
13. Updated docs to advise user that 'chmod +x ./bin/restore-database-backups.sh' may be needed in order to make the script executable

brecriffs commented:
@mlippert This is ready for you to take another look.

jaedoucette commented on README.md:
```diff
@@ -1,2 +1,85 @@
+
+### Database Backups
+
+Download sql backups from the Riff Learning backup location (ask for the URL if you don't have it), and place all .sql files in the folder called 'database-backups' in the root of this repo.
```
I'm concerned that we're downloading these files manually and in an un-encrypted format to our local machines. The mattermostDB has a lot of PII in it, and usually when handling something that sensitive in a research setting, I'd require data to be downloaded anew for each analysis in an encrypted format from a secure machine that requires 2-factor authentication to access; be decrypted into memory for analysis (so that unencrypted files aren't left lying around the machine); and only then analyzed. Ideally, the analysis script should erase the encrypted data from the machine when it's done running too. This isn't perfect, but it prevents a lot of the most dangerous behaviors by making sure that users don't leave PII data lying around unencrypted on regular machines. 

Probably this is okay for now, but adding that functionality should be a priority if we're going to use the metabase analysis often.

brecriffs replied:
Added a note in the README under 'Getting Started' regarding which ports are used on your local machine by the docker containers.

brecriffs replied:
> I left some comments on practices.
> 
> I needed to do some extra work besides what was in the documentation to run the metabase server. Specifically, I needed to stop my own mysql server that was running locally.
> 
> I was also not able to test the functionality of the metabase server at this time because the username/password combinations are not widely known. It might be a better idea to put them in a secrets file that we distribute over slack, so that it is easier for devs and analysts to find them as needed?

Added a note in the README under 'Getting Started' regarding which ports are used on your local machine by the docker containers.

jaedoucette commented on README.md:
````diff
@@ -1,2 +1,85 @@
+```
+make down
+```
+This command will stop and remove the mysql and metabase containers. The mysql volume containing the sql backups will
````
As a stop-gap, maybe a "make clean" command that users should run that erases the local copies of the mysql DB?

brecriffs replied:
My initial impression here was that you were suggesting make clean would run something like 'docker volume prune', after you run make down. Perhaps this command would also be useful to have in the Makefile? 

I have added 'make clean' to the Makefile which deletes all .sql files in the folder ./database-backups. I have added to the README a recommendation to run this command after restoring the backups to the mysql container.

jaedoucette replied:
The `docker volume prune` might also be nice to have as a make target, but the deletion one is what I was hoping for.

mlippert replied:
`docker volume prune` would be very bad to put in a target in the Makefile!
You would likely end up deleting volumes for containers completely unrelated to this project.
You have named volumes. Remove it/them explicitly.
`docker volume rm ba-mmdb-data`

A `make clean` which deletes the files which are manually obtained isn't great. When this is integrated w/ John's scripts used perhaps in a `make update-db` target, which downloads the latest backup automatically, then `make clean` should be removing that/those and probably the mysql volume it is restored into.

brecriffs replied:
Ah, agreed. Explicitly removing the relevant volumes makes more sense.

brecriffs replied:
I am not 100% clear on your second paragraph. What are you recommending I implement right now? My interpretation is: `make clean` should be run when you are done using the software, and it should delete the mysql backups as well as the mysql volume?

mlippert replied:
`clean` is a [common Makefile target](https://www.gnu.org/prep/standards/html_node/Standard-Targets.html) with a common somewhat expected meaning
> Delete all files in the current directory that are normally created by building the program. Also delete files in other directories if they are created by this makefile.

Basically it cleans up things created by the makefile. The `.sql` files are not created by the makefile and therefore the `clean` target shouldn't remove them. It would be fine if a different target removed them.

Is that clearer?

brecriffs replied:
Ok makes sense. You were suggesting, I think, that in the future, when a script downloads the sql backups, then` make clean` should remove them, since at that point, they are created/downloaded dynamically.

But for now, maybe `make clean` should:

1. Remove the metabase log file (metabase.db.trace.db)
2. run `docker volume rm ba-mmdb-data`

And another target, maybe `make remove-database-backups`, should delete all .sql files inside ./database-backups?

mlippert replied:
yep that's exactly what I was saying.

I personally don't think I'd bother w/ that additional target, but if you and John think it's useful, that is the way to do it.

brecriffs replied:
Ok makes sense. Thanks!

brecriffs replied:
It looks like the default behaviour when using docker-compose, is that the directory name is prepended to the volume name. So in this case, the volume is officially called `business-analytics_ba-mmdb-data`, and thus the command to delete will be `docker volume rm business-analytics_ba-mmdb-data`. Is it safe to say that the directory will always be `business-analytics`? Technically somebody could clone the repo into a folder with a different name.

mlippert replied:
What I'd suggest is that either you get the directory name yourself, or that you assume `business-analytics` for now and leave a comment that is what you're doing.

brecriffs replied:
Ok makes sense. I've addressed this in a new commit.

jaedoucette commented on README.md:
```diff
@@ -1,2 +1,85 @@
+persist locally however, so only make up or make up-detach is needed to start the software again.
+
+#### Restoring Database Backups
+Once the `make up` command has completed (about 2-3 minutes), open a new terminal tab and run `bin/restore-database-backups.sh`. It will restore databases from all .sql files in the folder ./database-backups. You may need to run `chmod +x ./bin/restore-database-backups.sh`, to make the file executable.
```
It might be preferable to automate this process? You can use short-circuit evaluation in make, so cmd1 && cmd2 should cause cmd2 to run only after the successful completion of cmd1.

jaedoucette replied:
Shouldn't the execute bit on `./bin/restore-database-backups.sh` be set in the repo now? I think that gets stored at the file level, so future users shouldn't need to do this. I could be mistaken though.

mlippert replied:
The issue is that just because the `make up` has successfully started all the containers doesn't mean that the mysql instance running in its container is ready for the restore to be started. Determining that is tricky.

brecriffs replied:
Yeah. There are solutions for detecting when the sql container is ready for a db backup to be restored. I guess if we determine that it's a priority for everything to happen with one command (which WOULD be cleaner), and if I get the go ahead, I can spend some time and implement that.

jaedoucette replied:
That makes more sense to me. Maybe try adding: `&& sleep 5s && ./bin/restore-database-backups.sh` as a hacky workaround? I like the single command build even if it's not completely correct.

mlippert replied:
you don't need that all in a single line, although putting it all in a single line means it all runs in the same subshell.
But in this case just run those all sequentially.
However, I really think this is a future optimization (you can figure out how to improve the base script to check that the container is started, then test if the container is ready, then do the restore)

brecriffs replied:
The swiftest way I found to do this was to run a looped grep on the mysql container logs to check for a line of text that signifies the container is ready. 

Ok so maybe I will hold off on this for now?

jaedoucette commented on README.md:
```diff
@@ -1,2 +1,85 @@
+
+## View the Dashboard
+
+Once you navigate to http://localhost:3000, you will be shown the metabase sign-in promt. Ask a developer for the login credentials. Once logged in, you will be presented with the metabase homepage. I have pinned the Mattermost dashboard to the
```
Maybe clearer to say: "Open a web-browser and navigate to http://localhost:3000"

brecriffs replied:
Made that change.

jaedoucette commented on README.md:
```diff
@@ -1,2 +1,85 @@
+
+## Making Changes in Metabase
+
+If you make any changes to a question or dashboard, remember to click SAVE. At this point, you will want to commit the
```
Consider here adding a screenshot with "SAVE" highlighted. Visuals are more likely to be noticed by readers, and losing work is frustrating if they miss this point.

mlippert replied:
I haven't looked into how one would add an image to a README.md. Do you know how @jaedoucette or @brecriffs ?

jaedoucette replied:
![Image description](link-to-image)

jaedoucette replied:
Doh! `![Image description](link-to-image)`

mlippert replied:
yeah but where would that image live, and how do you get it there? Ah you're suggesting committing the image perhaps in a docs folder or `docs/images` folder so the `link-to-image` would be `docs/images/image-name.jpg` or some such.

brecriffs replied:
Turns out the README supports ```<img src="docs/images/question-save-btn-preview.png" width="500"/>```. Added this.

Committed locally so it is not in the .gitignore. Would it be better to have this /docs folder only in the remote repo?

brecriffs replied:
Also, is it desirable to have screenshots accompanying every instance of 'click this'?

jaedoucette replied:
I tend to think so, but it depends on the hassle to get them and target audience. Here, I think the long run goal is for less technical people to be able to run this from the docs, in which case more pictures seems good.

mlippert replied:
> Committed locally so it is not in the .gitignore. Would it be better to have this /docs folder only in the remote repo?

@brecriffs I'm not sure what you mean here.

Also width=500? Does the syntax that John mentioned not work well for your image? sorta makes me wonder how big an image you created. I just looked, and my wondering was off base, nice image for the purpose, and a good size (resolution and file).

brecriffs replied:
I just meant that the added `/docs/images` folder was added locally, then committed, and I didn't add it to any gitignore. I guess I was wondering if you think it's desirable to have the /docs folder tracked? My thinking is, it might be nice to have the ability to add/edit some text/images in that folder locally.

brecriffs replied:
I just liked the syntax of the img tag I guess. The ability to specify width seems nice depending on the image. 500px looked nice in the docs on my laptop so I went with that.

mlippert replied:
If the images aren't committed and they are referenced by the readme (or other docs), how will those images be found to be displayed?
Certainly you can put files in that directory and not commit them without them being in the gitignore, it just can get tiresome seeing them listed as untracked all the time. You shouldn't be blindly committing every file that looks new or changed (using `git add` w/o specifying what to add), you should be committing files deliberately.

So what files are you thinking you'd want in docs/images locally but not committed? I can't think of any.

brecriffs replied:
My thought was, there might not be any sense in have the /docs/images folder locally, so if there was a way to have that folder only in the remote branch, to be referenced by the readme, then that could be desirable.

jaedoucette commented on bin/restore-database-backups.sh:
```diff
@@ -0,0 +1,15 @@
+
+function restoreDatabaseBackup() {
+   echo "Processing $1"
+   cat "$1" | docker exec -i ba-mysql /usr/bin/mysql --user=root --password=root
```
Also concerning that the PII-containing DB has default credentials, and that the credentials are documented in our repo. Ideally we'd prompt the user for this information, or else store the DB in an encrypted format.

mlippert replied:
This is only for the database running locally on the dev's machine, and to be honest this is an issue throughout, and not one I think we're going to be tackling any time soon. Since those databases aren't exposed to the internet, we've considered that "good enough".

brecriffs replied:
Yeah I think knowing the credentials only means that if you were to clone our repo, run the containers, run this script to restore some mysql backups, then you could access that mysql container. But the mysql container would only contain database backups that you had access to anyway.

jaedoucette replied:
Right. Brec's point makes sense to me. I think in the longer run we're going to want safer data handling procedures, but the credentials here aren't the right fix.

jaedoucette commented on README.md:
```diff
@@ -1,2 +1,85 @@
+
+### Database Backups
+
+Download sql backups from the Riff Learning backup location (ask for the URL if you don't have it), and place all .sql files in the folder called 'database-backups' in the root of this repo.
```
Another issue here: currently we download gzipped versions of the SQL db. consider adding instructions to `gzip -d *` in that directory, or something more detailed.

brecriffs replied:
Updated README to more accurately describe the process. 

1. Download backups
2. Unzip using gzip command
3. Place in ./database-backups

jaedoucette commented on README.md:
````diff
@@ -1,2 +1,85 @@
+The first step is to clone this repository. The following commands are defined in a Makefile, and can be run as follows:
+
+```
+make up
````
I am unable to run`make up`. I receive the output below. I do not appear to have any active docker containers. It looks like there's an error binding to the necessary ports:

`Starting Riff Business Analytics.
Creating network "business-analytics_default" with the default driver
Creating ba-mysql ... error

ERROR: for ba-mysql  Cannot start service ba-mmdb: driver failed programming external connectivity on endpoint ba-mysql (f568f539792f3a4beec3c188dd97ae70bed0f36cf3790e467635bff7762846c0): Error starting userland proxy: listen tcp 0.0.0.0:3306: bind: address already in use

ERROR: for ba-mmdb  Cannot start service ba-mmdb: driver failed programming external connectivity on endpoint ba-mysql (f568f539792f3a4beec3c188dd97ae70bed0f36cf3790e467635bff7762846c0): Error starting userland proxy: listen tcp 0.0.0.0:3306: bind: address already in use
ERROR: Encountered errors while bringing up the project.
Makefile:15: recipe for target 'up' failed
make: *** [up] Error 1`

jaedoucette replied:
This error is caused because the ba-mysql server attempts to bind to the default port used by mysql. Since I already had a mysql server active locally, this caused the build to fail.

The docs should either instruct users to shutdown any active mysql instances (e.g. `#mysqld stop`), or the `docker-compose` should bind to a different port that is less likely to be in use.

jaedoucette replied:
After stopping my local mysql instance, I was able to successfully run `make up`.

brecriffs replied:
Added this to the documentation under 'Getting Started'.

jaedoucette reviewed (commented):
I left some comments on practices.

I needed to do some extra work besides what was in the documentation to run the metabase server. Specifically, I needed to stop my own mysql server that was running locally.

I was also not able to test the functionality of the metabase server at this time because the username/password combinations are not widely known. It might be a better idea to put them in a secrets file that we distribute over slack, so that it is easier for devs and analysts to find them as needed?

mlippert commented:
> I needed to do some extra work besides what was in the documentation to run the metabase server. Specifically, I needed to stop my own mysql server that was running locally.
That is an interesting point. and I can see going both ways. We don't _have_ to publish the mysql container's port to localhost, which is what was causing you this issue, however it can be nice to have available. So perhaps just documenting that nothing should be using the standard mysql port (I don't know it off the top of my head) is sufficient.
> 
> I was also not able to test the functionality of the metabase server at this time because the username/password combinations are not widely known. It might be a better idea to put them in a secrets file that we distribute over slack, so that it is easier for devs and analysts to find them as needed?
For now there aren't that many analysts/devs so that this isn't much of a hardship. And if we eventually don't need to commit the metabase db, because a new one is created, then knowing the user/pswd isn't necessary. Meantime referencing a slack post w/ this specific information is probably a good idea.

brecriffs committed:
Update README.md

Added note regarding the ports that the docker containers use

Changed words of instructions on navigating to port 3000 on localhost

brecriffs committed:
Added 'make clean' command to Makefile

make clean runs 'rm ./database-backups/*.sql'

Updated README to reflect this new command's functionality

brecriffs committed:
Added instructions for unzipping sql backups

brecriffs committed:
Added image of save btn to README

Added folder docs/images to store the image

mlippert reviewed (approved):
OK, I think there's more to be done, in particular I think the sql for the questions should be checked in, in their own files (I pushed an example for channel data if you want to see what I mean).

But this now meets it's initial goals as far as I'm concerned. And I can see why @brecriffs wants to check in the metabase database (although we need to work on how that is done).
It makes it easy to extract the information we'll want.
I still believe that we should have instructions for creating it from scratch eventually, and that all question's sql should be in their own `.sql` files (which will help for the instructions for setting up metabase from scratch), as well as help track changes to the queries.

brecriffs commented:
> OK, I think there's more to be done, in particular I think the sql for the questions should be checked in, in their own files (I pushed an example for channel data if you want to see what I mean).
> 
> But this now meets it's initial goals as far as I'm concerned. And I can see why @brecriffs wants to check in the metabase database (although we need to work on how that is done).
> It makes it easy to extract the information we'll want.
> I still believe that we should have instructions for creating it from scratch eventually, and that all question's sql should be in their own `.sql` files (which will help for the instructions for setting up metabase from scratch), as well as help track changes to the queries.

Checking in the mysql queries separately is a great idea. I am also curious to see how the metabase db handles branch merging. For example, if we both add a new question or edit a question on different branches, then merge into master.

mlippert commented:
> I am also curious to see how the metabase db handles branch merging. For example, if we both add a new question or edit a question on different branches, then merge into master.

Yeah, that's not going to work **at all**! The database is binary file and therefore isn't diffable or mergable. That's perhaps the biggest problem I had with checking in the database.

@jaedoucette if you're OK with calling this branch good, would you approve the PR, and then I'll merge it. Thanks.

jaedoucette reviewed (approved):
Approved.

brecriffs committed:
Added clean and remove-database-backups targets to Makefile

Also added to the readme to reflect these changes

mlippert commented on Makefile:
```diff
@@ -1,15 +1,23 @@
+	echo "         and removes the mysql volume"    								 ; \
+	echo ""																													 ; \
+	echo "- down: run docker-compose down"                           ; \
+	echo ""																													 ; \
```
The command separator and line continuation that are at the end of the lines above are now all over the place.
I prefer them neatly lined up myself, but they should be either neatly lined up OR just follow the command on the line w/ a single space if you don't want to line them up.

mlippert commented on Makefile:
```diff
@@ -1,15 +1,23 @@
 #
 # Makefile to run the business analytics software using metabase
 #
+#||: means or true, and prevents the target from crashing if there is an error
```
This is a good comment because now I understand what you were doing below.
However what you probably should be doing for this situation is not forcing a 0 return, but instead just [ignoring error returns](https://www.gnu.org/software/make/manual/html_node/Errors.html) from the command which you do in a Makefile with a `-` prefix.
